src/main/java/teammates/common/util/Logger.java

```diff
@@ -18,38 +28,98 @@ public static Logger getLogger() {
      */
-    public void severe(String msg) {
-        log.severe(msg);
+    public void severe(String message) {
+        errorLog.severe(formatLogMessage(message, "ERROR"));
+    }
+
+    private String formatLogMessage(String message, String severity) {
+        if (Config.isDevServer()) {
+            return formatLogMessageForHumanDisplay(message);
+        }
+        return formatLogMessageForCloudLogging(message, severity);
+    }
+
+    private String formatLogMessageForHumanDisplay(String message) {
+        StringBuilder prefix = new StringBuilder();
+
+        StackTraceElement source = getLoggerSource();
+        if (source != null) {
+            prefix.append(source.getClassName()).append(':')
+                    .append(source.getMethodName()).append(':')
+                    .append(source.getLineNumber()).append(':');
+        }
+        prefix.append(' ');
+
+        if (RequestTracer.getRequestId() == null) {
+            return prefix.toString() + message;
+        }
+        return prefix.toString() + "[" + RequestTracer.getRequestId() + "] " + message;
+    }
+
+    private String formatLogMessageForCloudLogging(String message, String severity) {
+        return JsonUtils.toCompactJson(getBaseCloudLoggingPayload(message, severity));
+    }
+
+    private Map<String, Object> getBaseCloudLoggingPayload(String message, String severity) {
+        Map<String, Object> payload = new HashMap<>();
+        payload.put("message", message);
+        payload.put("severity", severity);
+
+        StackTraceElement source = getLoggerSource();
+        if (source != null) {
+            Map<String, Object> sourceLocation = new HashMap<>();
+            sourceLocation.put("file", source.getClassName());
+            sourceLocation.put("line", source.getLineNumber());
+            sourceLocation.put("function", source.getMethodName());
+
+            payload.put("logging.googleapis.com/sourceLocation", sourceLocation);
+        }
+
+        if (RequestTracer.getRequestId() != null) {
+            String[] traceAndSpan = RequestTracer.getRequestId().split("/", 2);
+            payload.put("logging.googleapis.com/trace", "projects/" + Config.APP_ID + "/traces/" + traceAndSpan[0]);
+            if (traceAndSpan.length == 2) {
+                payload.put("logging.googleapis.com/spanId", traceAndSpan[1].split(";")[0]);
+            }
+        }
+
+        return payload;
+    }
+
+    private StackTraceElement getLoggerSource() {
+        StackTraceElement[] stes = Thread.currentThread().getStackTrace();
+        for (int i = 0; i < stes.length; i++) {
+            StackTraceElement ste = stes[i];
+            if (ste.getClassName().equals(Logger.class.getName()) && i + 1 < stes.length) {
+                StackTraceElement nextSte = stes[i + 1];
+                if (!nextSte.getClassName().equals(Logger.class.getName())) {
+                    return nextSte;
+                }
+            }
+        }
+        return null;
```

Comment:
We should not be calling this when we are at the bottom of the stack ideally - seems odd that we have to check for `null` everywhere from the return of this method.

Reply:
Any idea to figure out "when we are at the bottom of the stack"?

Reply:
An alternative might be to just throw an exception since it's not something that should be expected e.g. this function should not be called when there is nothing on the stack

Reply:
I don't really like this class interrupting program flow. This is an internal system logger; the business logic of the system performs just fine even if this class does not exist.
If an exception is to be thrown here only to be caught and handled by the same class, it is effectively the same as the current implementation.

Reply:
Yes that is a fair point